### DISCUSSION TITLE
update_and_rebuild_libmesh error: fparser_ad.hh: No such file or directory

### DISCUSSION THREAD

**Discussion by aruybalid** [idaholab/moose#20673] [Q&A Getting Started] [answered]
Hi,

I am trying to install MOOSE on the INL HPC and get the following error upon running  ./update_and_rebuild_libmesh.sh. 

```pre
_make[1]: Entering directory `/home/ruybandr/projects/moose/libmesh/build'
  CXX      src/apps/fparser_parse_opt-fparser_parse.o
In file included from ../src/apps/fparser_parse.C:23:
./include/libmesh/parsed_function.h:34:10: fatal error: libmesh/fparser_ad.hh: No such file or directory
   34 | #include "libmesh/fparser_ad.hh"
      |          ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated._
```

Any idea what might be wrong?

Andre

**Comment by GiudGiud** [comment]
Hello

Weird. I have fparser_ad.hh under libmesh/include/libmesh.
Did petsc install ok?
What modules did you load?
Are you following the INL HPC instructions?

Guillaume

**Reply by aruybalid** [reply]
petsc installed OK

**Comment by aruybalid** [comment]
Hi Guillaume,

I am following instructions here: https://mooseframework.inl.gov/getting_started/installation/inl_hpc_install_moose.html

Currently Loaded Modules:
  1) gcc/9.3.0-gcc-4.8.5-twls   2) openmpi/4.0.5_ucx1.9   3) python/3.7-anaconda-2019.10   4) cmake/3.22.3-gcc-9.3.0-ngra

**Reply by GiudGiud** [reply]
You should have a different set of modules at this point, from having ran this:
`module load use.moose moose-dev`
(if you do run this command now, make sure to module purge first AND to re-install petsc)

Still wont explain the libmesh failure..

Are you on a recent moose commit? What does `git show` in moose/ say

**Reply by aruybalid** [reply]
Hi Guillaume,

I am using a moose fork from Topher Matthews:

```
commit e0d62f95a9f713f2f8c168162a0a7a72ff2938a9
Author: Topher Matthews <cmatthews@lanl.gov>
Date:   Mon Mar 28 17:05:04 2022 -0600

    Fix global limit calculation. It's still not perfect, and mising tiles at the extremes may still hit errors, but it should be good for most

diff --git a/modules/tensor_mechanics/include/materials/LAROMANCEStressUpdateBase.h b/modules/tensor_mechanics/include/materials/LAROMANCEStres
index 0cc8d41..3282533 100644
--- a/modules/tensor_mechanics/include/materials/LAROMANCEStressUpdateBase.h
+++ b/modules/tensor_mechanics/include/materials/LAROMANCEStressUpdateBase.h
@@ -126,16 +126,6 @@ protected:
                             const unsigned int t,
                             const unsigned int tt,
                             const unsigned int in_index);
-  /**
-   * Method to check input values against applicability windows set by ROM data set.
-   * @param input Input value
-   * @param vector for upper and lower WindowFailure enums indicating what to do if input is outside
-   * of limits
-   * @param global_limits Vector of lower and upper global limits of the input
-   */
-  void checkInputWindow(const GenericReal<is_ad> & input,
-                        const std::vector<WindowFailure> behavior,
-                        const std::vector<Real> & global_limits);
 
   /**
    * Convert the input variables into the form expected by the ROM Legendre polynomials to have a
@@ -419,8 +409,8 @@ protected:
 
   /*
    * Material specific orientations of tiling
-   * variables
-   * @return Vector declaring tiling orientation
+   * variables. Indexing is partition, then input
+   * @return Vector of a vector declaring tiling orientation
    */
   virtual std::vector<std::vector<unsigned int>> getTilings()
   {
@@ -431,8 +421,8 @@ protected:

```
I will load the proper modules and try to install again.

**Comment by aruybalid** [comment]
Hi Guillaume,

I just pulled origin/next and tried to install petsc (successful) and libmesh, but get the same error:

Here is the version of moose I now have and from which I tried to install libmesh:

```
ruybandr@sawtooth2:~/projects/moose/scripts$git show
commit 824960ff17002d85f032f6ac80037c26bbd12942 (HEAD -> next, origin/next, andre/next)
Merge: 6e5116d03d 26b0c8f844
Author: Logan Harbour <loganhharbour@gmail.com>
Date:   Tue Mar 29 10:31:16 2022 -0600

    Merge pull request #20336 from milljm/20335-generic-hash-names
    
    Return a hash for a given library
```

**Reply by GiudGiud** [reply]
Thanks for ruling this out. 
What does `find . -name fparser*` return in moose/libmesh?
and `echo $LIBMESH_DIR` ?

**Reply by aruybalid** [reply]
Thanks for the support:

`find . -name fparser*` finds 

```
./libmesh/tests/fparser
./libmesh/include/libmesh/fparser_ad.hh
./libmesh/include/libmesh/fparser.hh
./libmesh/bin/fparser_parse-devel
./libmesh/bin/fparser_parse-dbg
./libmesh/bin/fparser_parse-opt
./libmesh/src/apps/fparser_parse.C
./libmesh/installed/include/libmesh/fparser_ad.hh
./libmesh/installed/include/libmesh/fparser.hh
./libmesh/installed/bin/fparser_parse-devel
./libmesh/installed/bin/fparser_parse-dbg
./libmesh/installed/bin/fparser_parse-opt
./libmesh/m4/fparser.m4
./libmesh/contrib/fparser
./libmesh/contrib/fparser/fparser_gmpint.hh
./libmesh/contrib/fparser/fparser_ad.cc
./libmesh/contrib/fparser/fparser_ad.hh
./libmesh/contrib/fparser/fparser.cc
./libmesh/contrib/fparser/fparser_mpfr.hh
./libmesh/contrib/fparser/fparser.hh
./libmesh/contrib/fparser/docs/fparser.html
./libmesh/build/tests/fparser
./libmesh/build/src/apps/.deps/fparser_parse_opt-fparser_parse.Po
./libmesh/build/src/apps/.deps/fparser_parse_dbg-fparser_parse.Po
./libmesh/build/src/apps/.deps/fparser_parse_devel-fparser_parse.Po
./libmesh/build/contrib/fparser
./python/mms/fparser.py
```

So the file is in ./libmesh/include/libmesh/fparser_ad.hh, as you also reported, but it seems to be looking in libmesh/?

`echo $LIBMESH_DIR` gives:

/home/ruybandr/projects/moose/libmesh

**Reply by GiudGiud** [reply]
Yeah it is just there. I dont get it. Did you modify the `LIBRARY_PATH` ? 
Is this a fresh install ?
@milljm you seen this before?

**Reply by aruybalid** [reply]
I purged modules and loaded the modules as on the instruction page and per your suggestion earlier. Then I pulled the moose/next branch just before running the update_and_rebuild_libmesh.sh script (did petsc before libmesh). I did not alter LIBRARY_PATH.

**Reply by GiudGiud** [reply]
@loganharbour @cticenhour any idea on this?

**Reply by aruybalid** [reply]
I have also just tried it with idaholab/devel, but ran into the same problem. LIBMESH_DIR is: /home/ruybandr/projects/moose/libmesh, which is exported to $PATH within bashrc. Would it be useful to remove that directory from $PATH and start over to let the installation scripts set the directories as needed?

**Reply by GiudGiud** [reply]
I dont think you need LIBMESH_DIR exported to your PATH. Why was this done?

**Reply by aruybalid** [reply]
I miscommunicated. Excuse me for the confusion.  I have not added it to $PATH. I just set the LIBMESH_DIR to be /home/ruybandr/projects/moose/libmesh using `export` inside .bashrc. Is that the right thing to do?

**Reply by GiudGiud** [reply]
You shouldn’t have to do this for the update and install script I think. Can you try without?
Is that path correct btw? What does ‘ls $LIBMESH_DIR’ return?

**Reply by milljm** [reply]
While using our vetted version of libMesh (submodule libmesh) it is not necessary to set LIBMESH_DIR. If you find there is a foreign LIBMESH_DIR being set to something not-of-your-making (say from one of our Conda installs), the best way to deal with that, is to `unset LIBMESH_DIR`, and then build your own (which, is what you are trying to do).

As for the error... no, I have not encountered this before. I am still piecing things together...

**Reply by aruybalid** [reply]
`ls $LIBMESH_DIR` returns:
```
aclocal.m4  bootstrap  ChangeLog  configure.ac  doc       include    lib          Makefile.am  README              share
AUTHORS     build      CITATION   contrib       etc       INSTALL    m4           Makefile.in  README.md           src
bin         build-aux  configure  COPYING       examples  installed  Make.common  NEWS         reference_elements  tests
```
I will `unset LIBMESH_DIR` and try to run the update_and_rebuild_libmesh.sh script again. Stand by..

**Reply by aruybalid** [reply]
still the same error: Why is it looking for a file here:
```
./include/libmesh/parsed_function.h:34:10: fatal error: libmesh/fparser_ad.hh: No such file or directory
 #include "libmesh/fparser_ad.hh"
          ^~~~~~~~~~~~~~~~~~~~~~~
```
But the file it seems to be missing is there, inside `moose/libmesh/include/libmesh`. Since the file that specifies to include it: parsed_function.h, is in the same directory as the file it intends to include, shouldn't it point to `fparser_ad.hh` instead of `libmesh/fparser_ad.hh`?

**Reply by GiudGiud** [reply]
Can you please attach the configure.log file for libmesh? Should be in the libmesh/build directory.
Maybe that ll give us a clue

**Comment by aruybalid** [comment]
[config.log](https://github.com/idaholab/moose/files/8401112/config.log)

**Comment by GiudGiud** [comment, marked as answer]
Closing this.
We managed to get a working build on /scratch, it's still unclear what happens on home/. I ll keep an eye out see if others are affected by this
